Woo Connect Insights: Table Component

psealock opened the pull request:
### Create a component for tabular data
A dumb component that renders a table. 

Examples of usage 
![screen shot 2017-05-31 at 12 52 00 pm](https://cloud.githubusercontent.com/assets/1160732/26628845/fa7d3bd2-45ff-11e7-82b7-dc3657eff68f.png)

![screen shot 2017-06-07 at 7 55 44 pm](https://user-images.githubusercontent.com/1922453/26867969-5eba9552-4bbb-11e7-9869-a851fd40628b.png)

### Requirements
* The first cell should handle long names with a white fade out
* contents of the cells can be strings, nodes, or any other component

psealock commented:
### Progress so far
![screen shot 2017-06-08 at 2 07 42 pm](https://user-images.githubusercontent.com/1922453/26909299-e7141d04-4c53-11e7-86a9-2c60d7332c4b.png)

I'm headed down the `<table>` route instead of flexbox. I have to dynamically set the width of the first column to be able to layout the white fade out. 

At the moment, I'm setting the width individually on rows, which is bad. Especially if I have resize handlers on screen resize.

psealock commented:
### To Test
Since devdocs doesn't pull in extensions the only way to play around is to stick this somewhere:

```js
import Example from 'woocommerce/components/table/docs/example';

...

return (
    <Example/>
);
```
### API
I wanted to make this component as dumb as possible. In other words, simply display whats passed via props and remove as much logic as possible. 
```js
<Table values={ [ 'First', 'Second', 'Third' ] }>
    <Row values={ [ 1, 2, 3 ] }/>
    <Row values={ [ 4, 5, 6 ] }/>
</Table>
```

greenafrican commented on client/extensions/woocommerce/components/table/table-row/index.js:
```diff
@@ -0,0 +1,37 @@
+				if ( index === 0 ) {
+					return (
+						<El className={ 'table-row__cell is-title-cell' } key={ value + index }>
+							<div className="table-row__cell-title" >{ value }</div>
```
Add in a `title={ value.toString() }` property to the div so that you get the full value in a tooltip:
![screen shot 2017-06-08 at 2 02 29 pm](https://user-images.githubusercontent.com/1160732/26927711-487e0038-4c53-11e7-9a76-61422eb7465b.png)

greenafrican commented:
This is great @psealock!

Other future considerations:
- pass a format per column (eg. '%' or date format)
- pass a component as a cell value (eg. sparkline)
- sort rows by a column

ryelle commented:
Hi @psealock, it looks like we were both doing this at the same time 🙂  I started a table component for my work on the orders list in #14871, specifically https://github.com/Automattic/wp-calypso/pull/14871/commits/0f2b825ef00da352ccaaa515fde4793c4134fa9f. I think we had slightly different design aims, so maybe we can merge the two approaches.

I think our main code-difference is that I added specific elements for `TableHeading` and `TableItem`. I focused on keeping the table straightforward and accessible, by following [the tips in this article](http://webaim.org/techniques/tables/data), specifically:

- **Identify Row and Column Headers**: using `th` for any "label" information, column names of course,  but this also means that header cells can show up in the middle of a row (the order name in the case of orders list)
- **Associate the Data Cells with the Appropriate Headers**: using the `scope` attribute to define what axis of table the heading refers to. I defaulted it to col, since that seemed the most likely, but in the case of the order's name, that `th` should have `scope="row"`.

Maybe by adding an `Item`/`TableItem` with `isHeader`, we can bring in the accessibility options. Let me know how you'd like to collaborate on this — I can add directly to this PR, or just give feedback.

ryelle commented:
I noticed we're on very different timezones, so I just took a bit to try a merged approach that uses both styles, calling yours `compact` and mine `has-border`... I think this is now overly complicated, but I need to run for the day. Here's a branch of your branch with my code, if you want to work on it during my overnight :) https://github.com/Automattic/wp-calypso/commit/d9b3d1ccdc8c5c1b692fd88247019128d5f2d61c — if you don't get to it, I'll try to clean it up tomorrow.

psealock commented:
@ryelle Thats great! Hopefully we can get a single component to handle each of our needs.

>I think we had slightly different design aims, so maybe we can merge the two approaches.

Besides some simple CSS rules like `line-height` and `color`, the designs diverge in a few things. Lines separating rows and the alignment inside cells, for instance. @kellychoffman @raoulwegat, Do you think we could come up with a unified table theme for these types of things?

The most glaring difference is the Store Stats requirement for the 1st cell in each row to be largest and do a white-fade-out-thing if the string is too long:

![screen shot 2017-06-09 at 10 38 41 am](https://user-images.githubusercontent.com/1922453/26953656-e33dfb30-4cff-11e7-8d3c-e57b439bfe30.png)

Going further, the 1st cell needs to be `text-align: left` and the others `text-align: right` and pulled all the way to the right. We could retain this use case via an optional prop, perhaps `isNumericalData` ?

@ryelle the article brings up things I hadn't considered yet, thanks. Today I will take a look at your approach and put forward a proposal.

kellychoffman commented:
> Lines separating rows

I think we should carry this over as it would help with readability here, especially on the narrower view points. My suggested design from a previous p2 thread:

<img width="313" alt="underline" src="https://user-images.githubusercontent.com/5835847/26954143-5a96d078-4c7c-11e7-80ee-c74b610826a6.png">

> alignment inside cells

I think for prices we can both agree to align those to the right. I don't think we should align numbers to the right and alpha text to the left, it feels off. The means the Quantity and Total columns above should be left aligned.

> The most glaring difference is the Store Stats requirement for the 1st cell in each row to be largest and do a white-fade-out-thing if the string is too long

I don't think we need to standardize the length of cells. The fade out thing makes me nervous to use in this context as it could potentially hide important data. Are you only using it for the Name cell?

psealock commented:
>I don't think we need to standardize the length of cells. The fade out thing makes me nervous to use in this context as it could potentially hide important data. Are you only using it for the Name cell?

I am only doing it for the Name cell (first one in the row). The idea was to retain the look of Site Stats so Store and Site stats feel to be one and the same. 

![screen shot 2017-06-09 at 11 23 32 am](https://user-images.githubusercontent.com/1922453/26954769-27ef3ea0-4d06-11e7-92c1-8ed1eadb6616.png)

@kellychoffman A departure from this look might insert a visual distinction between Site and Store stats. What do you think?

kellychoffman commented:
> A departure from this look might insert a visual distinction between Site and Store stats. What do you think?

True. I think you should keep it. I believe for our Orders table, we should have a min width and horizontally scroll. I think it would be okay for us to diverge in this context - our 'full' size table horizontally scrolls and your 'compact' table utilizes fades. What do you think?

psealock commented:
>I think it would be okay for us to diverge in this context

Cool, @ryelle has already created a `compact` prop to change the layout.

psealock commented:
### Update
* Added a README
* Updated implementation and styles:

![table](https://user-images.githubusercontent.com/1922453/27060531-adeaa8ae-5031-11e7-80e9-dba8b94ce8a9.gif)

### A Note on Styles
Our two main use cases are covered here. Two other style cases came up in this conversation ([underlines](https://github.com/Automattic/wp-calypso/pull/14892#issuecomment-307252964) and dropped columns based on media queries) which are not addressed. Each sub component allows for inclusion of classes, so I think it will be best to address these cases in the javascript/CSS consuming these components.

Thanks to @ryelle for your help with this one.

kellychoffman commented:
Looks good! Just one thing on the tables with checkboxes. 

<img width="638" alt="screen shot 2017-06-12 at 7 02 43 pm" src="https://user-images.githubusercontent.com/5835847/27061523-e65c1ea6-4fa1-11e7-9df3-e37640692c00.png">

<img width="637" alt="screen shot 2017-06-12 at 7 01 54 pm" src="https://user-images.githubusercontent.com/5835847/27061528-ecaa954e-4fa1-11e7-8dae-fa834f80ee83.png">

There should only be a 16px (compact) / 24px (expanded) space between them. Like:
<img width="173" alt="screen shot 2017-06-12 at 7 07 57 pm" src="https://user-images.githubusercontent.com/5835847/27061606-a787b798-4fa2-11e7-942b-de05e210aa8a.png">
<img width="187" alt="screen shot 2017-06-12 at 7 08 04 pm" src="https://user-images.githubusercontent.com/5835847/27061607-a78904cc-4fa2-11e7-91c1-733a1df00bf0.png">

psealock commented:
Nice catch @kellychoffman. I fixed the padding to reflect the values you mentioned. For the "expanded" table I should note that the table is responsive with `width: 100%` and default column widths, so that space will shrink/grow.

I just had a look at some [wireframes](https://github.com/Automattic/wp-calypso/issues/13654#issuecomment-306613311) and it looks like there will be a need to `text-align: right` for currency values. I added a prop to control this.

cc @ryelle

psealock commented:
Also, when creating Woo charts we can place the label in the same cell as the checkbox to avoid this large space. This will allow clicking on the label to check/uncheck the box as well.

![screen shot 2017-06-13 at 3 19 32 pm](https://user-images.githubusercontent.com/1922453/27064511-e08e62b8-504b-11e7-83c2-e6c10086de27.png)
